### ISSUE TITLE
I consider sniparser harmful

### ISSUE THREAD

**Issue by Mygod** [shadowsocks/shadowsocks-libev#2627]
SNI parser was introduced in #809. While this feature might be useful for routers, I argue that this is the **wrong** way to tackle the DNS rerouting problem -- an ideal solution should be setting up a DNS server/relay, like `LocalDnsServer` in shadowsocks-android. Just a quick comparison...

Advantage of using sniparser:

1. No need to set up a DNS server. (though if you have gone through the trouble of getting shadowsocks-libev on your router, you might as well just do the extra step)

Disadvantage:

1. User cannot specify DNS server for HTTP/TLS traffic.
2. User might not even be aware of this hack, causing confusions. (#1503, #1505, #1507)
3. Only works properly for HTTP/TLS traffic, given that sniparser was implemented correctly (#1876, #1877). If a client uses HTTP/TLS mixed with other traffic, the result will be surprising.
4. If non HTTP/TLS traffic wishes to use this feature, one needs to write their protocol such that it can be identified by sniparser which is not and should not be standardized. (folks at IETF will probably be displeased to see this)

Proposal: Remove sniparser entirely. If this is too radical, add an option and disable it by default.

**Comment by madeye** [comment]
I'm okay with disabling it by default. A pull request?

**Comment by Mygod** [comment]
Why not remove it entirely and make everyone's life much easier? :)

**Comment by Mygod** [comment]
Also having a proxy tool trying to do DNS hijacking is definitely against what KISS is about...

**Comment by madeye** [comment]
After we find a better way to implement GFW list in shadowsocks-android, I'm okay to remove it entirely from this project.

Currently, it's only used by shadowsocks-android.

**Comment by Mygod** [comment]
> Currently, it's only used by shadowsocks-android.

Are you sure about this? Because whenever I look at local.c (even after removing `__ANDROID__` blocks) my sanity drains very quickly. 🤢

**Comment by madeye** [comment]
Yeah, I mean the only third-party software I know that depends on SNI feature is shadowsocks-android...

**Comment by Mygod** [comment]
I don't think shadowsocks-android needs SNI now that we have a proper DNS relay since https://github.com/shadowsocks/shadowsocks-android/issues/2346, correct?

**Comment by madeye** [comment]
Without SNI parser, GFW List (Proxy Only) mode won't work.

**Comment by Mygod** [comment]
Why? If requests come in as `SOCKS5_ATYP_DOMAIN` we will not be able to handle it either (unless we send an additional DNS probe), no?

**Comment by madeye** [comment]
tun2socks only sends SOCKS5_ATYP_IP to ss-local, which doesn't know the domain name for each request.

**Comment by Mygod** [comment]
Okay let me rephrase my previous comment... If ss-local gets a `SOCKS5_ATYP_DOMAIN`, it should already handle it correctly without an SNI parser; if ss-local gets a `SOCKS5_ATYP_IPV?` with HTTP/TLS payload, I consider this a bug of socks5 client rather than ss-local attempting to workaround that bug.

**Comment by Mygod** [comment]
In other words, let us consider one thing at a time. The point I am trying to make in this issue is that SNI parser is under no circumstances useful.

**Comment by madeye** [comment]
I agree with you on disabling SNI parser by default in shadowsocks-libev, which should not be handled here.

But for the first step, I suggest just disabling it and deprecate it in the next releases. After we solve the shadowsocks-android dependency issue, let's remove it entirely.

**Comment by Mygod** [comment]
No I am saying there is no dependency from shadowsocks-android. Removing it will not break a thing.

**Comment by studentmain** [comment]
Another problem is encrypted SNI.

**Comment by yzou** [comment]
> Okay let me rephrase my previous comment... If ss-local gets a `SOCKS5_ATYP_DOMAIN`, it should already handle it correctly without an SNI parser; if ss-local gets a `SOCKS5_ATYP_IPV?` with HTTP/TLS payload, I consider this a bug of socks5 client rather than ss-local attempting to workaround that bug.

This is not a bug of socks5 client, Firefox even has an option to resolve dns locally or remotely through socks5 server.

**Comment by Mygod** [comment]
And one of them is the wrong behavior.

**Comment by yzou** [comment]
On Android, the App doesn't know it's going through socks5, it just resolves domains and sends out requests normally.

**Comment by Mygod** [comment]
I don't have time to explain it to you.
